### PR TITLE
separate AL nonlinear params

### PR DESCRIPTION
**Pull request by Huangzizhou**

### CODE REVIEW

File: src/polyfem/solver/ALSolver.cpp [hunk 1]
```diff
@@ -5,28 +5,24 @@
...
 		  scaling(scaling),
 		  max_al_weight(max_al_weight),
 		  eta_tol(eta_tol),
-		  max_solver_iter(max_solver_iter),
 		  update_barrier_stiffness(update_barrier_stiffness)
 	{
 	}
 
-	void ALSolver::solve(NLProblem &nl_problem, Eigen::MatrixXd &sol)
+	void ALSolver::solve_al(std::shared_ptr<NLSolver> nl_solver,NLProblem &nl_problem, Eigen::MatrixXd &sol)
```
Comment (teseoch) [inline src/polyfem/solver/ALSolver.cpp:25]: why moving it from ctr to func?

File: src/polyfem/solver/ALSolver.cpp [hunk 2]
```diff
@@ -86,6 +81,19 @@ namespace polyfem::solver
 		}
 		nl_problem.line_search_end();
 		nl_solver->max_iterations() = iters;
+	}
+
+	void ALSolver::solve_al(std::shared_ptr<NLSolver> nl_solver,NLProblem &nl_problem, Eigen::MatrixXd &sol)
```
Comment (teseoch) [inline src/polyfem/solver/ALSolver.cpp:86]: is this duplicated?

File: src/polyfem/state/StateInit.cpp [hunk 3]
```diff
@@ -232,6 +232,9 @@ namespace polyfem
 		else
 			polysolve::linear::Solver::select_valid_solver(args_in["solver"]["adjoint_linear"], logger());
 
+		if (args_in["solver"]["augmented_lagrangian"]["nonlinear"].is_null())
```
Comment (teseoch) [inline src/polyfem/state/StateInit.cpp:235]: i think we need merge patch here to add any not set params in the al

File: src/polyfem/solver/ALSolver.cpp (hunk 1 above)
Comment (Huangzizhou) [reply src/polyfem/solver/ALSolver.cpp:25]: because we need a different solver for the last solve in ALSolver -- the newton reduced solve.

**Review by teseoch — APPROVED** [review]

**Pull request merged**